File: backend/app/schemas/diagram.py
from pydantic import BaseModel
from typing import List


class Position(BaseModel):
    x: float
    y: float


class Node(BaseModel):
    id: str
    type: str
    label: str
    position: Position
    parent: str | None = None


class Edge(BaseModel):
    id: str
    source: str
    target: str


class Group(BaseModel):
    id: str
    label: str
    position: Position
    width: float
    height: float


class Diagram(BaseModel):
    id: str
    nodes: list[Node]
    edges: list[Edge]
    groups: list[Group]
    metadata: dict = {}

    

File: frontend/src/types/diagram.ts
export type Position = {
  x: number
  y: number
}

export type DiagramNode = {
  id: string
  type: string
  label: string
  position: Position
  parent: string | null
}

export type DiagramEdge = {
  id: string
  source: string
  target: string
  label?: string
  routePoints?: Position[]
  arrowType?: 'none' | 'forward' | 'backward' | 'both'
  sourceSide?: 'auto' | 'top' | 'right' | 'bottom' | 'left'
  targetSide?: 'auto' | 'top' | 'right' | 'bottom' | 'left'
  style?: {
    lineStyle?: 'solid' | 'dashed' | 'dotted'
  }
}


export type DiagramGroup = {
  id: string
  label: string
  position: Position
  width: number
  height: number
}

export type Diagram = {
  id: string
  nodes: DiagramNode[]
  edges: DiagramEdge[]
  groups: DiagramGroup[]
  metadata: Record<string, unknown>
}
